src/Npgsql/NpgsqlTransaction.cs

```diff
@@ -41,7 +41,13 @@ public sealed class NpgsqlTransaction : DbTransaction
         /// <summary>
         /// If true, the transaction has been committed/rolled back, but not disposed.
         /// </summary>
-        internal bool IsCompleted => _connector.TransactionStatus == TransactionStatus.Idle;
+        internal bool IsCompleted => isCompleted || _connector.TransactionStatus == TransactionStatus.Idle;
+
+        /// <summary>
+        /// If true, the transaction has been committed/rolled back, but not disposed.
+        /// </summary>
+        /// <remarks>Useful for multiplexing, when on transaction commit/rollback the connector might be returned to the pool</remarks>
+        bool isCompleted;
```

Comment:
I might be misunderstanding what you're trying to do, but I'm not sure this mechanism is going to work. Each instance of NpgsqlTransaction belongs to its connector, and is recycled (Init is called when a new transaction starts). This means that if the connector is returned to the pool and then immediately given out again, and a new transaction is started, isCompleted will be set to false.

The only proper way to fix these race conditions is never do anything to the connector after we return it to the pool. I think this means redoing EndUserAction and scope unbinding to link them together, as mentioned above.

If I'm correct and you agree with this direction, I propose we leave this PR open for now and not do these changes (since the above will affect everything anyway). Except for the `connector.Connection = null` fix above, which is unrelated.

Reply:
> I might be misunderstanding what you're trying to do, but I'm not sure this mechanism is going to work. Each instance of NpgsqlTransaction belongs to its connector, and is recycled (Init is called when a new transaction starts). This means that if the connector is returned to the pool and then immediately given out again, and a new transaction is started, isCompleted will be set to false.

This is something, I've come to realize today, while I was doing more testing 😢.

> The only proper way to fix these race conditions is never do anything to the connector after we return it to the pool. I think this means redoing EndUserAction and scope unbinding to link them together, as mentioned above.

Maybe. Or maybe we can go the way @YohDeadfall [proposed](https://github.com/npgsql/npgsql/issues/2964#issuecomment-696023486) for the `NpgsqlDataReader`.

> If I'm correct and you agree with this direction, I propose we leave this PR open for now and not do these changes (since the above will affect everything anyway). Except for the `connector.Connection = null` fix above, which is unrelated.

You are correct and I do agree - multiplexing is hard (even though it's cool), and this fix can wait.

Reply:
>> The only proper way to fix these race conditions is never do anything to the connector after we return it to the pool. I think this means redoing EndUserAction and scope unbinding to link them together, as mentioned above.
>
> Maybe. Or maybe we can go the way @YohDeadfall proposed for the NpgsqlDataReader.

You mean we should consider to not recycle NpgsqlTransaction anymore? Sure, but that's the easy way out :rofl: One of the reasons Npgsql has such good perf (it's one of the reasons .NET Core is so high on TechEmpower) is because we try to pay attention to every little detail etc.

I've opened #3260 to track looking into this - unless you're very interested I can take this on. 

Note: I do think we should do something here before releasing 5.0. At minimum we can avoid starting a user action altogether if in multiplexing.